### PR TITLE
[Merged by Bors] - feat: Chinese remainder for ZMod.

### PR DESCRIPTION
**Pull request by PatrickMassot**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Ideal/Operations.lean [hunk 1]
```diff
@@ -632,13 +632,19 @@ theorem finset_inf_span_singleton {ι : Type*} (s : Finset ι) (I : ι → R)
...
-    ⨅ i, Ideal.span ({I i} : Set R) = Ideal.span {∏ i, I i} := by
+    ⨅ i, span ({I i} : Set R) = span {∏ i, I i} := by
   rw [← Finset.inf_univ_eq_iInf, finset_inf_span_singleton]
   rwa [Finset.coe_univ, Set.pairwise_univ]
 #align ideal.infi_span_singleton Ideal.iInf_span_singleton
 
+theorem Ideal.iInf_span_singleton_natCast {R : Type*} [CommRing R] {ι : Type*} [Fintype ι]
+    {I : ι → ℕ} (hI : ∀ (i j : ι), i ≠ j → Nat.Coprime (I i) (I j)) :
+    ⨅ (i : ι), Ideal.span {(I i : R)} = Ideal.span {((∏ i : ι, I i : ℕ) : R)} := by
+  rw [Ideal.iInf_span_singleton, Nat.cast_prod]
```
Comment (alreadydone) [inline Mathlib/RingTheory/Ideal/Operations.lean:645]:
```suggestion
theorem iInf_span_singleton_natCast {R : Type*} [CommRing R] {ι : Type*} [Fintype ι]
    {I : ι → ℕ} (hI : ∀ (i j : ι), i ≠ j → Nat.Coprime (I i) (I j)) :
    ⨅ (i : ι), span {(I i : R)} = span {((∏ i : ι, I i : ℕ) : R)} := by
  rw [iInf_span_singleton, Nat.cast_prod]
```
(untested)

File: Mathlib/Data/ZMod/Quotient.lean [hunk 2]
```diff
@@ -67,6 +67,22 @@ def quotientSpanEquivZMod (a : ℤ) : ℤ ⧸ Ideal.span ({a} : Set ℤ) ≃+* Z
...
+/-- The **Chinese remainder theorem**, elementary version for `ZMod`. See also
+`Mathlib.Data.ZMod.Basic` for versions involving only two numbers. -/
+def ZMod.prodEquivPi {ι : Type*} [Fintype ι] (a : ι → ℕ)
+    (coprime : ∀ i j, i ≠ j → Nat.Coprime (a i) (a j)) : ZMod (∏ i, a i) ≃+* ∀ i, ZMod (a i) :=
+  have : ∀ (i j : ι), i ≠ j → IsCoprime (span {(a i : ℤ)}) (span {(a j : ℤ)}) :=
+  fun i j h ↦ (isCoprime_span_singleton_iff _ _).mpr ((coprime i j h).cast (R := ℤ))
+  Int.quotientSpanNatEquivZMod _ |>.symm.trans <|
+  quotEquivOfEq (iInf_span_singleton_natCast (R := ℤ) coprime) |>.symm.trans <|
+  quotientInfRingEquivPiQuotient _ this |>.trans <|
+  RingEquiv.piCongrRight fun i ↦ Int.quotientSpanNatEquivZMod (a i)
```
Comment (alreadydone) [inline Mathlib/Data/ZMod/Quotient.lean:82]: You could write `(a i).Coprime (a j)` if you want (and elsewhere).

Is this "flat style" the new fashion? I saw that a new linter is [enforcing something like this](https://github.com/leanprover-community/mathlib4/pull/7286#discussion_r1341776263) but I think it's not desirable.

**Review by alreadydone — APPROVED** [review]

**Comment by alreadydone** [comment]
Oh I just saw this is WIP ... approved too early :)

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**